packages/eslint-plugin/src/rules/enum-utils/shared.ts

```diff
@@ -33,8 +48,36 @@ export function getEnumTypes(
   typeChecker: ts.TypeChecker,
   type: ts.Type,
 ): ts.Type[] {
-  return tsutils
-    .unionTypeParts(type)
-    .filter(subType => util.isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral))
-    .map(type => getBaseEnumType(typeChecker, type));
+  return getEnumLiterals(type).map(type => getBaseEnumType(typeChecker, type));
+}
+
+/**
+ * Returns the enum key that matches the given literal node, or null if none
+ * match. For example:
+ * ```ts
+ * enum Fruit {
+ *   Apple = 'apple',
+ *   Banana = 'banana',
+ * }
+ *
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'apple') --> 'Fruit.Apple'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'banana') --> 'Fruit.Banana'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'orange') --> null
+ * ```
+ */
+export function getEnumKeyForLiteral(
+  enumLiterals: ts.Type[],
+  literal: unknown,
+): string | null {
+  for (const enumLiteral of enumLiterals) {
+    // @ts-expect-error Not sure why `value` is not on `enumLiteral`.
+    if (enumLiteral.value === literal) {
+      const { symbol } = enumLiteral;
+
+      // @ts-expect-error Not sure why `parent` is not on `symbol`.
+      return `${symbol.parent.name}.${symbol.name}`;
```

Comment:
🤔 great question. I don't know. It must be part of the private API...

In the past we've had to augment TypeScript's types. See `packages/*/typescript/typescript.d.ts`. But this is very risky (what if they update their internals?) so we try not to.

Instead, you could go through the AST: access `symbol.valueDeclaration` and assert that it's an `EnumMember`. You'll then be encouraged by the type system to handle all the possible allowed cases (what if the member name is a literal like `'abc'`?).

Reply:
I think I got it right this time, and it's simpler than what you've suggested, thanks!